fix: infinite mixin recursion through class_alias

sdkawata opened the pull request:
- [x] Added or updated tests
- [ ] Documented user facing changes

Resolves #1883

**Changes**

In MixinPropertiesClassReflectionExtension, there's a mechanism to prevent infinite mixin recursion. However, within ReflectionHelper, because a new instance of MixinPropertiesClassReflectionExtension is created each time, it wasn't functioning correctly.
In https://github.com/larastan/larastan/pull/1839, an attempt was made to prevent infinite recursion where a class is mixin is applied to itself, but it failed to prevent infinite recursion when mixin was applied via class_alias. This PR addresses this issue by modifying the code to reuse MixinPropertiesClassReflectionExtension instances instead of creating new ones each time, thus effectively preventing recursion over a wider range of mixin scenarios.

**Breaking changes**

Probably No

calebdw commented:
Why would a class mixin an alias of itself?

sdkawata commented:
> Why would a class mixin an alias of itself?

`\Illuminate\Support\Facades\Redis` mixin `\Redis`.
https://github.com/larastan/larastan/blob/66db15f82209d971e9bcca99ea3b2cb031be9d52/stubs/common/Facades.stub#L8

And,  `\Redis` is an alias of `\Illuminate\Support\Facades\Redis`.
https://github.com/laravel/laravel/blob/3923e7f7c40368a5e78c1a33610191be8ad91e3b/config/app.php#L218

This recursion cause https://github.com/larastan/larastan/issues/1883

calebdw commented:
I'm fairly certain this is (or rather should be) mixing in the root Redis class/client from phpredis and not the alias:

https://github.com/laravel/framework/blob/5710fdb8bf36df501405c705f7d2ea4720d2b9d3/src/Illuminate/Redis/Connections/Connection.php#L18-L23

There's no reason for a class to mixin an alias to itself

sdkawata commented:
While I agree that this intended to be mixing the Redis class from phpredis,
this unintentionally be mixing an alias in the environment where phpredis is not installed,
and larastan crashes in such an environment.
This happens for laravel default installation (-7.0), thus this fix may benefit many users.

This PR not only fixes the issue of infinite mixin recursion through class_alias
but also prevents infinite mixin recursion in other scenarios (such as #1830).
Therefore, it may benefit users who might encounter similar issues under different circumstances.

canvural commented:
Hi,

I was suspecting this would not cause the errors you mentioned and I removed your changes for `ReflectionHelper` but kept the tests and tests still passed 🙂 Also this test might be more suitable to `Integration` folder.

So first please try to create the error in the CI then we can talk about the fix.

sdkawata committed:
add failing test reproducing #1883

sdkawata committed:
fix: infinite mixin recursion through class_alias

sdkawata commented:
> I was suspecting this would not cause the errors you mentioned and I removed your changes for ReflectionHelper but kept the tests and tests still passed 🙂 Also this test might be more suitable to Integration folder.

I moved the test to `Integration` folder and confirmed that the added test fails in CI,

canvural commented:
> and confirmed that the added test fails in CI,

It did not fail for the correct reason 😄 Looks like it's a memory limit issue in PHPUnit. You can try to increase the memory limit for PHPUnit or set it to `-1` And please just keep the failing test. I want to see the same segmentation fault error in the CI that was mentioned in #1883

calebdw commented:
Note that this is also just testing an unknown class on alias mixin---this is not testing the specific code that caused the  segfault:

https://github.com/larastan/larastan/issues/1883#issuecomment-1999647373

canvural commented:
Hi,

I just remembered that there is an undocumented PHPStan config option just for this case! It's called `mixinExcludeClasses` So you can add the following to your `phpstan.neon` file
```neon
mixinExcludeClasses:
	Redis
```

and it should solve the problem 👍🏽
